### PR TITLE
Bring some order to modules

### PR DESCRIPTION
**Pull request by jpakkane**
Our module API is a bit of a wild west. This is first step towards something better. Create a return value object that encapsulates the result of a module invocation. It contains two items

 - the object to return to Meson script
 - the list of new objects to append to build etc

The implementation is broken in many ways and many tests do not pass. The point of this is to be something to get the discussion rolling.

This also adds code that checks that the module does not alter build objects behind our back. All modules should be append only and return new things in the array, the interpreter takes care of putting them in their respective places. This way we have alteration logic in one centralized place.

### CODE REVIEW

File: mesonbuild/modules/modtest.py [hunk 1]
```diff
@@ -12,10 +12,14 @@
...
 # limitations under the License.
 
+from . import ModuleReturnValue
+
 class TestModule:
 
     def print_hello(self, state, args, kwargs):
         print('Hello from a Meson module')
+        rv = ModuleReturnValue(None, [])
+        return rv
```
Comment (ignatenkobrain) [inline mesonbuild/modules/modtest.py]: why not to return it directly?

**Comment by ignatenkobrain** [comment]
looks good, though I don't really understand concept of this ModuleReturnValue... =(

**Comment by TingPing** [comment]
Basically a `namedtuple` as I see it, more readable.

**Comment by jpakkane** [comment]
This allows modules to create targets that are not returned as return values to the Meson script. Thus they don't need to poke the state dictionaries by themselves and we can now guard against that with the hashcheck.

**Comment by kyrylo-shpytsya-prism** [comment]
Does this potentially pave a road towards third party modules? If so, how does it fit with the "makes it possible to port Meson's implementation to a different language" part mentioned [here](https://github.com/mesonbuild/meson/wiki/FAQ#why-is-meson-not-just-a-python-module-so-i-could-code-my-build-setup-in-python)?
My scenario is an implementation of a certain framework which is to be integrated with various build systems (currently cmake and meson). By integration I mean that the user just unpacks the archive into a subdirectory of his/her project, and adds **minimal** amount of boilerplate code into his/her build-system files to enable the framework do it's magic. The magic involves execution of an external program, creating a static library target and returning some data such as e.g. paths to include files and name of the produced static library target.
In cmake, I can do "include" of a cmake file from the framework and call a function defined there. Moreover, that function can be called multiple times.
Currently, I do not see a way to achieve such a functionality in meson.
My dream syntax would be something like

>r=import("<my framework>").do_magic("arg1", "arg2", "file-name-arg"...)
>executable(..., link_with: [r.to_link], include_directories: [r.to_include])

Related to this (please tell me if a should open separate feature requests):

- add ability to implement modules in meson's own language
- add support for user defined functions
- add "parse json" function (or maybe a method of string class) to be able to conveniently return complex data from external executables.

**Comment by jpakkane** [comment]
We do not aim to support a stable API interface for modules. Furthermore Meson language is not, nor will it ever be, Turing complete so defining your own functions will never be supported.

We aim to support these sort of use cases via the rewriter tool. Currently you can add and remove sources to a target programmatically from the command line. The goal is to extend that to support more operations.

**Comment by kshpytsya** [comment]
@jpakkane Let me elaborate on my intended use case:

Typical workflow:
* download the framework as an archive and unpack it into subfolder of project source tree, or do "git submodule" mantras instead
* install prerequisites (think `pip`)
* create configuration files for the framework (e.g. json, yaml and some other DSLs)
* add **small** (being a perfectionist, I would struggle for literally every last character short of sacrificing readability)  amount of "code" to the project description file (CMakeLists.txt, meson.build) that:
     * "imports" the framework, 
     * "invokes" framework supplying it the configuration files by a file name prefix (or a name of a dedicated directory)
     * uses the following artefacts returned from the "invocation":
        * name (or some reference object) of a static library target, created by the framework (containing multiple generated source files, e.g. in C++), for the parent target to link to
        * include path (or some reference object) to add to the parent target
        * list of source (maybe yet non-existent) files that need to be added to parent target (the intent here is to force the parent target to have those specifically named files and to avoid duplication, i.e. user should neither need to manually add files to parent target if configuration says they have to be added, nor manually remove them, if they are no longer referenced by the configuration)
* run build system setup tool (cmake, meson) which would, among the other typical activities, execute generation tool from the framework that would generate all the necessary source files
* (optionally) load the project in IDE supported by the build system, i.e. VS, XCode, Codelite, Codeblocks and be able to navigate source files, including the generated ones, in a familiar manner
* build project using chosen backend, e.g. ninja
* †edit configuration files
* build using backend or IDE. The configuration file(s) change should trigger generator tool execution which may or may not change the list of generated source files constituting the static library and the list of source files that are to be part of the parent project. Ideally, if IDE is used, it should notice change of the project/"solution" file and reload. Regretfully at least Codelite and Codeblocks on Linux fail at this (requiring manual user reload or the project), but I vaguely recall Visual Studio doing it (been wa-a-ay too long ago).
* rinse, repeat †

My goals:
- simplicity and foolproof-ness;
- support for multiple build systems: currently I target cmake and meson;
- once initially set up (using cmake/meson), no changes to framework configuration should ever require user to manually re-setup (i.e. rerun cmake/meson) or execute some other tool to bring build system up to date
- strict adherence to "do not repeat yourself" principle (see explanation on list of source files to "force upon" the parent project)
- dynamic set of generated and "forced" source files as seen by the build system and IDE. While it would be possible to simplify a lot of things by having a single generated source file (which would possibly use "#include" to include other source files, generated and "forced"), this would open an even uglier can of worms. 
- ability to "invoke" framework multiple times, i.e. have, within the build project, multiple framework configurations producing multiple static libraries which would be linked to either different or the same parent target

I have been able to achieve my goals with cmake to degree that the "code" added to CMakeLists.txt is not uglier that the rest of typical cmake.

Currently, for CMake, the framework supplies a cmake file that has to be "included" and which defines a function that takes configuration file(s) name (prefix). The function uses `execute_process` to run the generator which generates source files and a cmake file that is included by the function right after the generation. This generated cmake file defines the static library target and uses `configure_file` trick to specify list of files that would cause regeneration of build. These files are all the configuration files (only generator knows their exact list) and the generator itself (to facilitate framework development). The invocation artefacts are returned via variables whose names are supplied in function call. 

With meson, I have successfully employed similar approach using `run_command` to run generator, then doing `subdir` to include the generated meson and the same `configure_file` trick to specify build regeneration dependencies. The idea with json was to avoid generation of meson.build and instead doing the creation of the target in the meson.build provided by the framework using the data returned by the generator via stdout in json format.  This approach seems cleaner to me than generating a meson.build.

Due to the lack of functions in meson, I fail to see how I can achieve multiple invocation goal cleanly. Currently I pass "invocation" parameters (i.e. name (prefix) of configuration file(s)) via variable and do "subdir(framework)". This, of course, can be done only once due to 'Tried to enter directory "...", which has already been visited.' error.

Two possible solutions that would work but severely lack elegance are:
* use symlinks to clone the framework to multiple directories. This is non-portable and an can wreak havoc with version control.
* process multiple configurations in a single place in the build system by having a single framework "invocation" process multiple configurations

Since I don't know intended functionality of `rewriter`, I cannot tell if it would allow me to achieve my desired workflow.

**Comment by kshpytsya** [comment]
@jpakkane Yet another solution that appears balanced to me ("beauty is in the eye of the beholder", of course) is to implement a "source/include meson build code from a pipe" function (which, as I have implemented, is a mixture of existing "run_command", a piece of "subdir" and some error handling). I think we would agree that the use case I have described in my previous comment cannot currently be implemented in meson (please correct me if I am wrong). So, to make it possible, some changes to meson are necessary. To me it seems that the "source from pipe" solution is minimally invasive to meson both code-wise (my implementation is a mere 24 lines) and concept-wise (i.e. not introducing new data types or Turing-completeness to the core meson).  You can clearly state in documentation that this is an escape hatch and whatever other scary words you would chose, but such things are sometimes necessary for some idealistic (no offense intended) systems to be practical in non-standard situations.
I would gladly create a pull request should you find the idea acceptable.

**Comment by jpakkane** [comment]
We already have builtin support for exactly what you are trying to achieve: Subprojects and `declare_dependency`. See documentation [here](https://github.com/mesonbuild/meson/wiki/Dependencies), [here](https://github.com/mesonbuild/meson/wiki/Subprojects) and [here](https://github.com/mesonbuild/meson/wiki/Wrap%20dependency%20system%20manual).

**Comment by kshpytsya** [comment]
Is `sp.get_variable('i')` mentioned [here](https://github.com/mesonbuild/meson/wiki/Subprojects) the same as the standalone function [`get_variable`](https://github.com/mesonbuild/meson/wiki/Reference%20manual#get_variable)? How does one pass any kind of configuration to subprojects at the `subproject(...)` invocation in the meson.build?
Is it possible to invoke the same "subproject" multiple times? For me,
> sp1 = subproject('my_framework')
> sp2 = subproject('my_framework')
executed "my_framework" subproject just once.

**Comment by jpakkane** [comment]
> Is sp.get_variable('i') mentioned here the same as the standalone function get_variable?

Kind of. It gets variables from the subproject, plain `get_variable` gets them from the current project.

> How does one pass any kind of configuration to subprojects at the subproject(...)

Only by having project options in the subproject and passing default values to those: https://github.com/mesonbuild/meson/wiki/Reference-manual#subproject

> Is it possible to invoke the same "subproject" multiple times?

No. Each subproject is executed only once. Executing it again will return the same object as the first one. A subproject is only ever built once with one configuration per global project.

**Comment by kyrylo-shpytsya-prism** [comment]
Well, unless the subproject mechanism is extended to allow multiple invocations, it is not going to suit my needs.
Also, I haven't managed (after some trying and even adding defug output to meson itself) to pass some option from parent project to subproject.

I am trying the following:
> project('main', 'c')
> sp1 = subproject('framework', default_options: ['myopt=a'])

meson.build:
subprojects/framework/meson.build
> project('framework', 'c', default_options:['myopt=x'])
> message('framework:', get_option('myopt'))

And I get:

> Tried to access unknown option "framework:myopt"

Various combinations of replacing "myopt" by "framework:myopt" in `default_options` keyword argument, and adding it to the main `project(...)` didn't make a difference.

**Comment by jpakkane** [comment]
> Well, unless the subproject mechanism is extended to allow multiple invocations, it is not going to suit my needs

Why do you need to invoke it multiple times?

> Tried to access unknown option "framework:myopt"

You need to specify your options in the [options file](https://github.com/mesonbuild/meson/wiki/Build%20options).

**Comment by kyrylo-shpytsya-prism** [comment]
Multiple invocations are needed to satisfy the

> ability to "invoke" framework multiple times, i.e. have, within the build project, multiple framework configurations producing multiple static libraries which would be linked to either different or the same parent target

mentioned in my comment [above](https://github.com/mesonbuild/meson/pull/1278#issuecomment-271433572).

Imagine having a certain library that uses "the framework" to generate some code. Now the executable target that links with that library also uses "the framework" to generate some completely unrelated code. Or maybe it is a second library that also uses "the framework". The idea is to make it possible to keep all the details of the framework invocation local to each user (i.e. they should be in corresponding meson.build files).
Actually, use of subprojects for this seems a conceptual stretch to me, but "as long as it quacks like a duck"... but it doesn't. What I really need is something that quacks like "user implemented modules", be it in meson language itself (which, as far as I understand is not going to happen), or as a "plugin" to meson program (won't happen either, because you do not want to provide a stable module API, or even tie meson to python) , or as some other mechanism, one of the possibilities ("source pipe") for which I have suggested.
Having the end user paste a dozen of lines of boiler-plate code at each place "the framework" is to be "invoked" instead of a concise "import module and call function" just isn't satisfactory to me, esp. when I can easily achieve that in cmake, being it as ugly as it is.

**Comment by jpakkane** [comment]
What you are suggesting is to have multiple projects use the same subproject but with different options. If so this is dangerous, it is very easy to link two incompatible libraries in the same thing. As an example NPM allows you to have multiple different versions of the same dependency even. This is simple to do, but an absolute unfixable disaster once your project grows even a little bit.

It is highly recommended to have only one setup shared across all users or, if that is not feasible, then build the different version explicitly in the subproject and give them their own deps. Then you can do stuff like:

    sp = subproject('foobar')
    mydep = sp.get_variable('setup1')

in one project and

    sp = subproject('foobar')
    otherdep = sp.get_variable('setup2')

**Comment by kshpytsya** [comment]
Let me quote myself:

> Actually, use of subprojects for this seems a conceptual stretch to me, but "as long as it quacks like a duck"... but it doesn't. What I really need is something that quacks like "user implemented modules", be it in meson language itself (which, as far as I understand is not going to happen), or as a "plugin" to meson program (won't happen either, because you do not want to provide a stable module API, or even tie meson to python) , or as some other mechanism, one of the possibilities ("source pipe") for which I have suggested.
> Having the end user paste a dozen of lines of boiler-plate code at each place "the framework" is to be "invoked" instead of a concise "import module and call function" just isn't satisfactory to me, esp. when I can easily achieve that in cmake, being it as ugly as it is.

Basically, the idea is that modules are needed and so are 3rd partly modules, since it is unreasonable for you to implement modules for anything other than mainstream frameworks. 
Think of something like qt with it's `moc`: imagine forcing your users dealing with this without "qt" meson module. Think of anything else domain specific that involves similarly complex logic. Think of something proprietary.
If you do not provide support for 3rd party modules you are actually leaving those types of users overboard. Some of them were hoping to escape the ugliness of cmake but you give them no hope.

**Comment by jpakkane** [comment]
Almost all functionality in modules could be done with just plain Meson scripts. The original implementation of Moc was basically just that. Proprietary things should be possible without any help from the module system (though it may be a bit cumbersome). What we _do_ want is to have just _one_ implementation of all major frameworks (there are not that many of them), and we want them upstream and tested as part of our regular development cycle. We definitely do _not_ want to end up where there are five different implementations of some thing scattered around the Internet that are all broken in different ways.

**Comment by kshpytsya** [comment]
The key word is "cumbersome". My philosophy is to provide the team with a solution that minimizes cognitive load and possibilities for screw ups . Any copy-paste boilerplate code just does not fit in, as it is exactly the opposite of what "the framework" itself is trying to accomplish: one can do without code generation just wring plain C++ as well.
Regretfully this means that I will not be able to introduce meson to the workflow of my team as a cmake replacement. There is, of course, an theoretical option of using a forked version of meson but that is too much pain in a long run.
Kindly thank you for your attention.

**Pull request merged**